flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleUtils.java

```diff
@@ -163,6 +166,25 @@ static String binaryWriterWriteNull(
         return CodeGenUtils.binaryWriterWriteNull(indexTerm, writerTerm, logicalType);
     }
 
+    static String lengthExceedsPrecision(String strTerm, int precision) {
+        return methodCall(strTerm, "length") + " > " + precision;
+    }
+
+    /**
+     * For complex types such as {@link LogicalTypeRoot#ARRAY}, use {@link VarCharType#STRING_TYPE}
+     * to avoid trimming/padding when casting the inner elements to a {@link
+     * LogicalTypeRoot#VARCHAR} or {@link LogicalTypeRoot#CHAR}.
+     */
+    static LogicalType getLogicalTypeForInnerCast(LogicalType logicalType) {
+        return logicalType.is(LogicalTypeFamily.CHARACTER_STRING)
+                ? VarCharType.STRING_TYPE
+                : logicalType;
+    }
+
+    static boolean shouldPossiblyTrim(int precision) {
+        return precision < VarCharType.MAX_LENGTH;
+    }
```

Comment:
This class has only code gen methods, not predicates. Can we place these somewhere else?

Reply:
I could also move the `shouldPossiblyTrim()` (as the other I'll remove) to the `AnyToCharVarchar` class.